kernel/src/arch/x86_64/process.rs

```diff
@@ -102,93 +94,135 @@ lazy_static! {
                 };
                 process_logs.push(local_log_arc);
             }
-            return process_logs;
+            return process_logs;*/
+            unimplemented!("Need to get NodeReplicated from controller")
         }
 
-        let process_logs = {
-            let mut process_logs = Box::try_new(ArrayVec::new()).expect("Can't initialize process log vector.");
-            for _pid in 0..MAX_PROCESSES {
-                let log = Arc::try_new(
-                    Log::<<NrProcess<Ring3Process> as Dispatch>::WriteOperation>::new(LARGE_PAGE_SIZE),
-                )
-                .expect("Can't initialize process logs, out of memory.");
-                process_logs.push(log);
-            }
-            process_logs
-        };
+        // We want to allocate the logs in controller shared memory
+        use crate::memory::shmem_affinity::local_shmem_affinity;
+        let pcm = per_core_mem();
+        pcm.set_mem_affinity(local_shmem_affinity()).expect("Can't change affinity");
+        
+        // Want at least one replica...
+        let num_replicas =
+            NonZeroUsize::new(core::cmp::max(1, atopology::MACHINE_TOPOLOGY.num_nodes())).unwrap();
+        let mut processes = ArrayVec::new();
+
+        for _pid in 0..MAX_PROCESSES {
+            debug_assert_eq!(
+                *crate::environment::NODE_ID,
+                0,
+                "Expect initialization to happen on node 0."
+            );
 
-        #[cfg(feature = "rackscale")]
-        if crate::CMDLINE
-            .get()
-            .map_or(false, |c| c.mode == crate::cmdline::Mode::Controller)
-        {
-            // Reset mem allocator to use per core memory again
-            let pcm = per_core_mem();
-            pcm.set_mem_affinity(0 as atopology::NodeId).expect("Can't change affinity");
+            let process: Arc<NodeReplicated<NrProcess<Ring3Process>>> = Arc::try_new(
+                NodeReplicated::new(num_replicas, |afc: AffinityChange| {
+                    let pcm = kcb::per_core_mem();
+                    match afc {
+                        AffinityChange::Replica(r) => {
+                            pcm.set_mem_affinity(mid_to_shmem_affinity(r)).expect("Can't change affinity");
+                        }
+                        AffinityChange::Revert(_orig) => {
+                            pcm.set_mem_affinity(local_shmem_affinity()).expect("Can't set affinity")
```

Comment:
local_shmem_affinity() returns a value relative to where it's run.  Also - the memory allocator may have been non-shared originally. So we don't want to automatically revert to shmem regardless. I think we need to capture the actual previous value to do this correctly.